Ensure sorting imports in a modified file picks up the proper configuration

PeterJCLaw opened the pull request:
Fixes #4891.

- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Has a [news entry](https://github.com/Microsoft/vscode-python/tree/master/news) file (remember to thank yourself!)
- [x] Appropriate comments and documentation strings in the code
- [x] Has sufficient logging.
- [x] ~Has telemetry for enhancements.~
- [x] Unit tests & system/integration tests are added/updated
    Tests updated in `importSortProvider.unit.test.ts` and a regression test added to `extension.sort.test.ts`
- [ ] [Test plan](https://github.com/Microsoft/vscode-python/blob/master/.github/test_plan.md) is updated as appropriate
- [x] ~[`package-lock.json`](https://github.com/Microsoft/vscode-python/blob/master/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed)~
- [ ] The wiki is updated with any design decisions/details.

PeterJCLaw commented on src/client/providers/importSortProvider.ts:
```diff
@@ -38,41 +37,38 @@ export class SortImportsEditingProvider implements ISortImportsEditingProvider {
+            diffPatch = (await processExeService.exec([importScript].concat(args), spawnOptions)).stdout;
         }
+
+        return this.editorUtils.getWorkspaceEditsFromPatch(document.getText(), diffPatch, document.uri);
```
Suggest reviewing this part ignoring whitespace changes.

karthiknadig commented:
@PeterJCLaw Thank you for the PR. Can you look at the merge conflicts?

karrtikr commented on src/client/common/process/proc.ts:
```diff
@@ -112,6 +112,11 @@ export class ProcessService extends EventEmitter implements IProcessService {
             });
         });
 
+        if (options.input) {
```
Can you also add some tests for the changes in the file? Could be system or a unit test.

karrtikr replied:
We have tests for the method here https://github.com/karrtikr/vscode-python/blob/2b6a8f2d439fe9d5e66665ea46d8b690ac9b2c39/src/test/common/process/proc.observable.test.ts#L42.

PeterJCLaw replied:
Thanks for the pointer to the file :)

I think https://github.com/microsoft/vscode-python/pull/9128/commits/58afa140a716818c558d4b17cac4b382479b2ae9 does this.

karrtikr replied:
The tests you just added are failing on CI. Please look into it.

PeterJCLaw replied:
D'oh, 62cdfda83 should fix that.

Is there a way I can run these tests locally? I ran `npm run test:unittests`, but that didn't show any issues, so I'm guessing there's something I'm missing.

karrtikr replied:
Yes, you can follow the steps mentioned here.
https://github.com/microsoft/vscode-python/blob/master/CONTRIBUTING.md#running-system-tests

Unit tests only refers to files ending in `unit.test.ts`, which is why the tests you just added were not run.

karrtikr commented on src/client/common/process/types.ts:
```diff
@@ -28,6 +28,7 @@ export type ObservableExecutionResult<T extends string | Buffer> = {
 
 // tslint:disable-next-line:interface-name
 export type SpawnOptions = ChildProcessSpawnOptions & {
+    input?: string;
```
Also can you add a description for this as well?

PeterJCLaw replied:
Sure, would that be an inline docstring (do those exist in TypeScript?) or are there docs somewhere else?

karrtikr replied:
Yes, the inline one.

PeterJCLaw replied:
I think ef5066b6 does this.

DonJayamanne commented:
@karthiknadig /cc

PeterJCLaw commented:
@DonJayamanne thanks for the feedback.

The change to `execObservable` here was in order to keep the same interface between it in `exec`, though I agree it's a little contrived. I'd be happy to drop part of these changes and/or consider an alternative interface for `exec`.

Currently the `isort` provider uses `exec` and I'd hoped for a minimally invasive change by using the `SpawnOptions` to pass the `input`.

I had hoped that it would be generally useful to be able to pass arbitrary `input` to an `exec` and as that's the method used by the `isort` provider currently I think it would make keep that being the case.

An alternative change I had considered (but decided against) would be to have `exec` accept another (optional) parameter of `input: string`. Would that (and undoing the change to `execObservable`) be preferable? (This would allow the `isort` provide to keep using `exec`, which is simpler, while providing a generally useful enhancement)

PeterJCLaw commented:
(meta) @karrtikr thanks for your help so far, I'll pick up any other changes tomorrow evening (I'm in GMT).

PeterJCLaw commented on src/client/common/process/proc.ts:
```diff
@@ -112,6 +112,11 @@ export class ProcessService extends EventEmitter implements IProcessService {
             });
         });
 
+        if (options.input) {
```
Just had a thought about this while writing the docs for it -- should we have this do an explicit `null` or `undefined` check, rather than this implicit one? I'm assuming that (like in Python) this implicit check will be falsey for both `null` and empty string, yet the latter might end up being surprising to callers if they've explicitly provided a value which happens to be empty.

karthiknadig commented:
@PeterJCLaw we need to discus this with the team. Based on the decision we will let you know if we want to take this or go with @DonJayamanne 's suggestions.

PeterJCLaw commented:
@karthiknadig just wondering if there's an update on this?

karthiknadig commented:
@PeterJCLaw  Sorry, we haven't gotten to this yet. We will have a discussion on this item this sprint.

karthiknadig commented on src/client/providers/importSortProvider.ts:
```diff
@@ -38,41 +37,38 @@ export class SortImportsEditingProvider implements ISortImportsEditingProvider {
+        // We pass the content of the file to be sorted via stdin. This avoids
+        // saving the file (as well as a potential temporary file), but does
+        // mean that we need another way to tell `isort` where to look for
+        // configuration. We do that by setting the working direcotry to the
```
type-o `directory`

PeterJCLaw replied:
2254599d3; thanks!

karthiknadig commented:
@PeterJCLaw We discussed this issue. Can you update the PR based on Don's suggestion to use `execObservable` and use the `stdin` stream to pass the input. I understand this is a bigger change, but we believe this way would use the `exec*` APIs as intended. Let us know, if we can help in any way. We can take on this item, but it may be a while before we get to it since it is not on our priority list.

PeterJCLaw commented:
@karthiknadig thanks for the update.

A couple of questions, hopefully to speed up review next time around:
- Do you know of any examples of other places which use `execObservable` in a similar manner (roughly single invocation, passing stdin) which I could use as a reference?
- Do you have any thoughts on how best to avoid ending up duplicating the content of the `exec` while ensuring that the process handling is similarly robust (and stays in step)?
- Alternatively, if you'd rather it were duplicated, do you have patterns for keeping similar code in step across the codebase? (One I've seen used well is to have pairs of comments which suggest updating the other similar places when one is changed)

karthiknadig commented:
@DonJayamanne Can you provide guidance here on the usage?

PeterJCLaw commented:
Ok, I've updated this PR to move the process handling out into the provider. Unfortunately due to https://github.com/zeromq/zeromq.js/issues/386 I'm not actually able to test that this works as desired, though I have updated the tests.

My main uncertainty here is whether the provider layer (i.e: the code in `provideDocumentSortImportsEdits`) should be doing more explicit error handling and tidy-up of the process which is returned by `execObservable`. I _think_ from a read of `execObservable` that it doesn't need to, though I don't feel I understand the JavaScript/TypeScript patterns around this well enough to say for sure.

@karthiknadig would you be able to have a quick look at the code and let me know if there's more error handling/process cleanup which the code needs to do? (If there is, then pointers or examples towards doing that would also be much appreciated!)

PeterJCLaw commented:
> Unfortunately due to [zeromq/zeromq.js#386](https://github.com/zeromq/zeromq.js/issues/386) I'm not actually able to test that this works as desired, though I have updated the tests.

Downgrading `zeromq` to `6.0.0-beta.5` seems to let me work around that (and doesn't appear to cause any issues in the extension), so I've now checked that this works (and fixed some issues I'd missed).

@karthiknadig I think this is ready for review again now (please also see my comment above about whether this code needs to do more to tidyup the launched process).

PeterJCLaw commented:
@karthiknadig do you know if you're likely to have time to look at this soon?
I realise there's a test failure here, though it's limited to a single platform and I'm not sure either what could be causing it nor how to go about fixing it -- advice on either would be welcome.

karthiknadig commented:
@PeterJCLaw I will take a look at this sometime this week.

PeterJCLaw commented:
@karthiknadig I updated this for `master` recently, but it looks like it's still failing in the same way (that I don't really understand).

Any chance you'll be able to look at this soon?

karthiknadig commented:
@PeterJCLaw We are fixing some flaky tests. We will get to this will. I will take care of addressing this.

karthiknadig commented:
@PeterJCLaw Looks like `isort` does not pick up `.isort.cfg` when using `stdin` on windows.
```
C:\GIT\issues\s p\vscode-python\src\test\pythonFiles\sorting\withconfig>type "C:\GIT\issues\s p\vscode-python\src\test\pythonFiles\sorting\withconfig\original.py" | "C:\GIT\issues\s p\vscode-python\pythonFiles\lib\python\bin\isort" - --diff
```
If I pass the full path it seems to work as expected.
```
C:\GIT\issues\s p\vscode-python\src\test\pythonFiles\sorting\withconfig>"C:\GIT\issues\s p\vscode-python\pythonFiles\lib\python\bin\isort" original.py --diff
--- C:\GIT\issues\s p\vscode-python\src\test\pythonFiles\sorting\withconfig\original.py:before  2019-07-25 18:23:40.425993
+++ C:\GIT\issues\s p\vscode-python\src\test\pythonFiles\sorting\withconfig\original.py:after   2020-05-04 22:34:59.942799
@@ -1,3 +1,9 @@
-from third_party import (lib1, lib2, lib3,
-                         lib4, lib5, lib6,
-                         lib7, lib8, lib9)+from third_party import lib1
+from third_party import lib2
+from third_party import lib3
+from third_party import lib4
+from third_party import lib5
+from third_party import lib6
+from third_party import lib7
+from third_party import lib8
+from third_party import lib9
```
This is causing the test to fail. This should probably be filed as a bug on `isort`. This is a blocker to accept this change.

PeterJCLaw commented:
Thanks for getting back to me.

> @PeterJCLaw Looks like `isort` does not pick up `.isort.cfg` when using `stdin`.
> https://github.com/timothycrosley/isort/blob/5de8d50075f1dbe33a7e74172a6b6c27795a0c8d/isort/main.py#L596

Hrm, that's rather odd. What version of `isort` were you testing with? (And what version is the CI picking up?)

I can reproduce that issue (on Linux too) when using the latest development version of `isort` from GitHub, however using latest published version on PyPI (a fresh `pip install isort` on Python 3.6 on windows 10 gets version `4.3.21`) I'm unable to reproduce this.

The logs do seem to indicate that CI is picking up `4.3.21` though, so I'm not sure that this is the issue.

> This should probably be filed as a bug on `isort`.

There are actually a couple of issues with the latest isort development branch which look like they'll affect us -- I've raised them as https://github.com/timothycrosley/isort/issues/1188 and https://github.com/timothycrosley/isort/issues/1189.

---

Note that `isort` appears to now be developing a 5.x branch, rather than the current 4.3.x branch. That has dropped support for running on Python versions before 3.6 (though claims to still be able to parse & fix older versions) so I'm guessing there could be some more complications in the near future with that branch,

karthiknadig commented:
On windows 10 in am using ` 4.3.21`. CI is using the same version, since it uses requirements.txt that is pinned to that version.

PeterJCLaw commented:
Ah, I figured out why I couldn't reproduce this. I'm not sure what this means yet though.

For me, this works:
```
>isort - --diff < demo.py
--- :before     2020-05-06 00:09:28.178022
+++ :after      2020-05-06 00:09:28.178022
@@ -1 +1,3 @@
-import os, sys, collections
+import os
+import sys
+import collections

>
```

But this doesn't:
```
>type demo.py | isort - --diff

> 
```

I get a different diff if I change the config, so I think the config is a red herring here. I think the issue is going to be something around how pipes are handled.

I can also reproduce this on both Python 3.6.8 and 3.8.2 (on Windows 10). When I'm back at my other machine I'll try the pipe spelling on Linux, I wonder if it breaks there too? (I've only ever tried the `<` spelling there)

karthiknadig commented:
The test is failing on CI because it is not picking up `.isort.cfg`. The diff comes back empty, when it should be applying `force_single_line=True`. I see that the `cwd`  when spawning the process is set to the directory where the file is present. I don't see anything wrong there. I wonder why `isort` does not see the the config file.

karthiknadig commented:
@PeterJCLaw If we are always using diff and we are going to send file content over stdin in all cases, then may be we should update `sortImports.py` to use `isort.SortImports` instead of going through main. That API seems to do the right thing.

PeterJCLaw commented:
Ok, after some more digging, I don't think this relates to the configuration file at all. If I debug through `isort` in the failing case, it finds the configuration file just fine, but fails to read in the content of the python file after determining the encoding to use.

I think there are a couple of issues here:
- `isort` tries to `seek` within the incoming file, yet seeking within `stdin` isn't always possible in general and in Python doesn't seem to be as buffered might be assumed
- I _think_ there might be a Python bug where if you read of the end of a buffered stream and then try to `.seek(0)` you don't get back to the start, so even adding our own buffering wrapping (via `io.BufferedReader(...)`) doesn't solve this for short files, such as the one in this test

Notably, adding a few comment lines to the test file makes `isort` work for me.

I've reproduced these outside `isort` while being specific to Windows and small files being sent to Python via stdin. I _think_ there might be a Python bug here, which I've reported as https://bugs.python.org/issue40540.

Switching to using `isort.SortImports` directly can mitigate this as long as we pass `io.BufferedReader(sys.stdin.buffer)` as the file, though my testing shows that even that may not be sufficient for small files.
This might also have other impacts as the reader would then be an `IO[str]` rather than an `IO[bytes]`, which we would need to be sure didn't have any adverse effects.

karthiknadig commented:
For `isort.SortImports` we can pass the file contents directly using `file_contents` or dump the entire `stdin` into a stringIO and pass that as the file.
```
isort.SortImports(file_contents=text, show_diff=True)
```

PeterJCLaw commented:
> For `isort.SortImports` we can pass the file contents directly using `file_contents` or dump the entire `stdin` into a stringIO and pass that as the file.
> 
> ```
> isort.SortImports(file_contents=text, show_diff=True)
> ```

Yup, that sounds reasonable. I'll look at making that change this evening or tomorrow.

PeterJCLaw commented:
> > For `isort.SortImports` we can pass the file contents directly using `file_contents` or dump the entire `stdin` into a stringIO and pass that as the file.
> > ```
> > isort.SortImports(file_contents=text, show_diff=True)
> > ```
> 
> Yup, that sounds reasonable. I'll look at making that change this evening or tomorrow.

It's occurred to me that this would actually mean we'd either lose support for isort command line arguments or need to parse them (albeit using the `isort` functions) ourselves.

I'm currently exploring replacing `sys.stdin` with an `io.BytesIO`, which I think will also work but avoid that issue.

karrtikr commented on src/client/providers/importSortProvider.ts:
```diff
@@ -129,29 +102,74 @@ export class SortImportsEditingProvider implements ISortImportsEditingProvider {
+        const spawnOptions = {
+            token,
+            throwOnStdErr: true,
+            cwd: path.dirname(uri.fsPath)
```
Users expect the cwd to point to the workspace root instead: https://github.com/microsoft/vscode-python/issues/14254.
